woocommerce/woocommerce-ios
# Order Creation: Make product list searchable

Closes: #5685

## Description

This adds search to the product list (on the Add Product screen) during order creation. This performs a remote product search and displays the results in place.

## Changes

- Makes `SearchHeader` (already created and used in `FilterListSelector`) a public, reusable SwiftUI component and adds it to the `AddProductToOrder` view.
- Updates `AddProductToOrderViewModel` to handle the product search:
   - Adds a `searchProducts()` method to perform the remote search.
   - Updates `sync()` to either sync the full product list or search the product list, depending on whether a search term is passed.
   - Adds a published `searchTerm` and listens for changes. If a search term is entered, the results controller predicate is updated to fetch matching products and a sync is performed; if the search is cleared, the results controller predicate is reset (removing the search predicate) to display the full product list.
- Clears the search term when the Add Product screen is closed, using `onDisappear` in the `NewOrder` view. (This clears the search even if the modal is swiped to dismiss it, and doesn't clear the search when moving between the product list and a list of variations for a variable product.)
- Adds unit tests for searching the product list.

## Testing

1. Build and run the app.
2. Make sure Order Creation is enabled under Settings > Experimental Features.
3. Go to the Orders tab, tap the + button, and select "Create order."
4. On the New Order screen, tap the "Add product" button.
5. On the Add Product screen, confirm the product list loads as expected.
6. Enter a search term and confirm the expected search results appear. (Suggested tests: Try a search with no results, a search with some results, or a search with multiple pages of results.)
7. Select a variable product. Confirm the variations list appears, and when you navigate back to the product list your search is retained.
8. Close the product list (tapping "Close," swiping to dismiss, or selecting a product). Tap "Add product" again and confirm the full product list loads (your previous search is cleared).

## Screenshots


https://user-images.githubusercontent.com/8658164/153044711-13eed703-afa7-4c14-a520-dd8be8a074cb.mp4



## Submitter Checklist

Update release notes:
